### PR TITLE
Feat/default fund

### PR DESCRIPTION
**Pull request by ioanaroceanu2000**

### CODE REVIEW

File: contracts/storage/FeeCollectorStorage.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    mapping(address => uint256) internal defaultFund;
+    mapping(address => uint256) internal protocolFees;
+    bool internal defaultFundPaused;
+}
+
+contract FeeCollectorStorage is FeeCollectorStorageV1 {
+    // Reserve some storage for use in future versions, without creating conflicts
+    // with other inheritted contracts
+    // todo: figure out the slots amount
+    uint256[69] private __gap; // total storage = 100 slots, including structs
```
Comment (avniculae) [inline contracts/storage/FeeCollectorStorage.sol:17]: actual size and comment are not matching

File: deploy/11.feeCollector.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  console.log("proxyAddress ", proxyAddress);
+
+  const feeCollectorProxyInstance = (await hre.ethers.getContractAt(
+    "FeeCollector",
+    proxyAddress
+  )) as FeeCollector;
+
+  if (multisig !== deployer && multisig) {
+    // Transfer ownership
+    console.log(
```
Comment (avniculae) [inline deploy/11.feeCollector.ts:39]: console log after the actual tx

File: tasks/feeCollection.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+// Description:
+//   This task generates mutiple fee collection transactions
+//   if the --multisig flag is set. Otherwise, it executes the transaction.
+//
+// Example:
+//   ``npx hardhat collectFees --network mainnet --collect --distribute --transfer-protocol-fee --transfer-default-fund --multisig aUSDC_v1 aUSDC_v2``
+//
+task("collectFees", "Set pausability state")
```
Comment (avniculae) [inline tasks/feeCollection.ts:45]: update description

Comment (ioanaroceanu2000) [reply tasks/feeCollection.ts]: updated in [cbe77ae](https://github.com/Voltz-Protocol/voltz-core/pull/266/commits/cbe77ae66a92b88405421de1c72b1ee0227276a1)

File: contracts/storage/FeeCollectorStorage.sol (hunk 1 above)
Comment (ioanaroceanu2000) [reply contracts/storage/FeeCollectorStorage.sol]: I gave it 50 slots, buffer of 47 [cbe77ae](https://github.com/Voltz-Protocol/voltz-core/pull/266/commits/cbe77ae66a92b88405421de1c72b1ee0227276a1)

File: contracts/FeeCollector.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    function distributeFees(address asset) public override onlyOwner returns (uint256 defaultFundsDelta, uint256 protocolFeesDelta) {
+        // todo: check for roudups/downs of division
+        uint256 contractBalance = IERC20Minimal(asset).balanceOf(address(this));
+        if ( defaultFundPaused ) {
+            protocolFees[asset] += contractBalance;
+
+            protocolFeesDelta = contractBalance;
+        } else {
+            defaultFund[asset] += contractBalance/2;
+            protocolFees[asset] += contractBalance/2;
```
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/FeeCollector.sol]: added this to default fund [cbe77ae](https://github.com/Voltz-Protocol/voltz-core/pull/266/commits/cbe77ae66a92b88405421de1c72b1ee0227276a1)

File: contracts/FeeCollector.sol [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "./storage/FeeCollectorStorage.sol";
+import "./interfaces/IFeeCollector.sol";
+import "./interfaces/IERC20Minimal.sol";
+import "./core_libraries/SafeTransferLib.sol";
+
+/**
+todo: remove
```
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/FeeCollector.sol]: removed [cbe77ae](https://github.com/Voltz-Protocol/voltz-core/pull/266/commits/cbe77ae66a92b88405421de1c72b1ee0227276a1)

File: contracts/FeeCollector.sol [hunk 6]
```diff
@@ -45,18 +37,18 @@ contract FeeCollector is
 
     /// @inheritdoc IFeeCollector
     function distributeFees(address asset) public override onlyOwner returns (uint256 defaultFundsDelta, uint256 protocolFeesDelta) {
-        // todo: check for roudups/downs of division
         uint256 contractBalance = IERC20Minimal(asset).balanceOf(address(this));
+        uint256 distributionAmount = contractBalance > (protocolFees[asset] + defaultFund[asset]) ?
```
Comment (avniculae) [inline contracts/FeeCollector.sol:41]: contract balance should always be equal to protocol fees + default fund, right? therefore, function should revert if this is not the case, so I would just do `contractBalance - protocol fees - default fund`.

File: contracts/FeeCollector.sol [hunk 7]
```diff
@@ -45,18 +37,18 @@ contract FeeCollector is
...
+            contractBalance - protocolFees[asset] - defaultFund[asset] : 0;
         if ( defaultFundPaused ) {
-            protocolFees[asset] += contractBalance;
-
-            protocolFeesDelta = contractBalance;
+            protocolFeesDelta = distributionAmount;
```
Comment (avniculae) [inline contracts/FeeCollector.sol:44]: I would remove variable

**Review by avniculae — APPROVED** [review]

**Pull request merged**